photutils/psf/daogroup.py

```diff
@@ -0,0 +1,129 @@
+from astropy.table import Column, Table, vstack
+
+__all__ = ['daogroup']
+
+def daogroup(starlist, crit_separation):
+    """
+    This is an implementation of the DAOGROUP algorithm presented by
+    Stetson (1987).
+
+    DAOGROUP divides an entire starlist into sets of distinct, self-contained
+    groups of mutually overlapping stars.
+
+    GROUP accepts as input a list of stars and determines which stars are
+    close enough to be capable of adversely influencing each others' profile
+    fits.
+
+    Parameters
+    ----------
+    starlist : `~astropy.table.Table` or array-like
+        List of stars positions.
+        If `~astropy.table.Table`, columns should be named 'x_0' and 'y_0'.
+        TODO: If array-like, it should be either (x_0, y_0).
+    crit_separation : float or int
+        Distance, in units of pixels, such that any two stars separated by
+        less than this distance will be placed in the same group.
+
+    Returns
+    -------
+    group_starlist : list of `~astropy.table.Table`
+        Each `~astropy.table.Table` in the list corresponds to a group of
+        mutually overlapping starts.
+
+    Notes
+    -----
+    Assuming the psf fwhm to be known, 'crit_separation' may be set to
+    k*fwhm, for some positive real k.
+
+    See
+    ---
+    `~daofind`
+    """
+
+    if not isinstance(crit_separation, (float, int)):
+        raise ValueError('crit_separation is expected to be either float or' +
+                         ' int. Received {}.'.format(type(crit_separation)))
+    if crit_separation < 0.0:
+        raise ValueError('crit_separation is expected to be a positive' +
+                         'real number.')
+
+    ## write a method that varifies whether the starlist given by the user
+    ## is valid
+
+    group_starlist = []
+    cstarlist = starlist.copy()
+
+    if 'id' not in cstarlist.colnames:
+        cstarlist.add_column(Column(name='id',
+                                    data=np.arange(len(cstarlist))))
+    
+    while len(cstarlist) is not 0:
+        init_group = _find_group(cstarlist[0], cstarlist, crit_separation)
+        assigned_stars_ids = np.intersect1d(cstarlist['id'], init_group['id'],
+                                            assume_unique=True)
+        cstarlist = _remove_stars(cstarlist, assigned_stars_ids)
+        n = 1
+        N = len(init_group)
+        while(n < N):    
+            tmp_group = _find_group(init_group[n], cstarlist, crit_separation)
+            if len(tmp_group) > 0:
+                assigned_stars_ids = np.intersect1d(cstarlist['id'],
+                                                    tmp_group['id'],
+                                                    assume_unique=True)
+                cstarlist = _remove_stars(cstarlist, assigned_stars_ids)
+                init_group = vstack([init_group, tmp_group])
+                N = len(init_group)
+            n = n + 1
+        group_starlist.append(init_group)
+    return group_starlist
```

Comment:
After looking at your code, I've been thinking that it looks more complex than it has to. I'll just sketch a possible modification below, I can work that out in more detail, of you think that might be a way to go:
```
# Your chec if 'id' exisits here...
starlist.add_column('id')
n = 1
while (starlist['id'] == 0).sum() > 0:
    # pick a star without id to start a new group
    init_star = starlist[np.where(starlist['id'] == 0)[0][0]]
    # change _find_group to return the index array itself and not a new table
    index = _find_group(init_star, starlist[starlist['id'] == 0]], crit_separation)
    starlist['id'][index] = n
    n = n + 1
# either return starlist, or don't return anything, since starlist is modified in place
```